### PR TITLE
jbrowse-cli add connection

### PR DESCRIPTION
**Pull request by peterkxie**
Command to add connection to JB2 config, help info below

```
Add a connection to a JBrowse 2 configuration

USAGE
  $ jbrowse add-connection DATADIRECTORY [LOCATION]

ARGUMENTS
  DATADIRECTORY  URL of data directory
                 For hub file, usually called hub.txt
                 For JBrowse 1, location of JB1 data directory similar 
                 to http://mysite.com/jbrowse/data/ 

  LOCATION       [default: .] Location of JBrowse installation. Defaults 
                 to .

OPTIONS
  -a, --assemblyName=assemblyName  Assembly name of the connection If 
                                   none, will default to the assembly in 
                                   your config file

  -c, --config=config              Any extra config settings to add to 
                                   connection in JSON object format, 
                                   such as 
                                   '{"uri":"url":"https://sample.com"}}'

  -f, --force                      Equivalent to `--skipCheck 
                                   --overwrite`

  -h, --help                       show CLI help

  -n, --name=name                  Name of the connection. Defaults to 
                                   connectionId if not provided

  -t, --type=type                  type of connection, ex. 
                                   JBrowse1Connection, 
                                   UCSCTrackHubConnection, custom

  --configLocation=configLocation  Write to a certain config.json file. 
                                   Defaults to location/config.json if 
                                   not specified

  --connectionId=connectionId      Id for the connection that must be 
                                   unique to JBrowse.  Defaults to 
                                   'connectionType-assemblyName-currentT
                                   ime'

  --overwrite                      Overwrites any existing connections 
                                   if same connection id

  --skipCheck                      Don't check whether or not the data 
                                   directory URL exists or if you are in 
                                   a JBrowse directory

EXAMPLES
  $ jbrowse add-connection http://mysite.com/jbrowse/data/
  $ jbrowse add-connection http://mysite.com/jbrowse/custom_data_folder/ 
  --type JBrowse1Connection
  $ jbrowse add-connection http://mysite.com/path/to/hub.txt 
  --assemblyName hg19
  $ jbrowse add-connection http://mysite.com/path/to/custom_hub_name.txt 
  --type UCSCTrackHubConnection --assemblyName hg19
  $ jbrowse add-connection http://mysite.com/path/to/custom --type 
  custom --config '{"uri":{"url":"https://mysite.com/path/to/custom"}}' 
  --assemblyName hg19
  $ jbrowse add-connection https://mysite.com/path/to/hub.txt 
  --connectionId newId --name newName
```

### CODE REVIEW

File: packages/jbrowse-cli/src/commands/add-connection.ts [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+  static args = [
+    {
+      name: 'dataDirectory',
+      required: true,
+      description: `URL of data directory\nFor hub file, usually called hub.txt\nFor JBrowse 1, location of JB1 data directory similar to http://mysite.com/jbrowse/data/ `,
+    },
+    {
+      name: 'location',
+      required: false,
+      description: 'Location of JBrowse installation. Defaults to .',
```
Comment (garrettjstevens) [inline packages/jbrowse-cli/src/commands/add-connection.ts:43]: oclif adds a description of the default if you use the "default" param, so technically you don't need the default in the description here.

File: packages/jbrowse-cli/src/commands/add-connection.ts [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+        }
+        if (!response || response.ok) return locationUrl.href
+      } catch (error) {
+        // ignore
+        this.error('Unable to fetch from URL', { exit: 90 })
+      }
+    }
+    return this.error(`Could not resolve to a URL: "${location}"`, {
+      exit: 100,
+    })
```
Comment (garrettjstevens) [inline packages/jbrowse-cli/src/commands/add-connection.ts:283]: Could be good to add the response code if the response is not ok, and the error message if the fetch throws an error.

File: packages/jbrowse-cli/src/commands/add-connection.ts [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+    }
+    return fsPromises.readFile(location, { encoding: 'utf8' })
+  }
+
+  determineConnectionType(url: string, config: string | undefined) {
+    if (path.basename(url) === 'hub.txt') return 'UCSCTrackHubConnection'
+    if (url.includes('jbrowse/data')) return 'JBrowse1Connection'
+    if (config && this.isValidJSON(config)) return 'custom'
+    return this.error(
+      `Unable to determine a specific connection from URL given.\nPlease specify a type with --type.\nIf you want a custom type, please provide the config object with --config`,
```
Comment (garrettjstevens) [inline packages/jbrowse-cli/src/commands/add-connection.ts:305]: Probably should use `os.EOL` instead of `\n` for Windows support.

Comment (cmdcolin) [reply packages/jbrowse-cli/src/commands/add-connection.ts]: Does a console log really have to worry about os.EOL?

**Pull request merged**